Src/Support/Google.Apis.Auth.Tests/OAuth2/ImpersonatedCredentialTests.cs

```diff
@@ -0,0 +1,230 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Auth.OAuth2.Flows;
+using Google.Apis.Auth.OAuth2.Responses;
+using Google.Apis.Json;
+using Google.Apis.Tests.Mocks;
+using Google.Apis.Util;
+using Moq;
+using System;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Google.Apis.Auth.Tests.OAuth2
+{
+    public class ImpersonatedCredentialTests
+    {
+        private IClock clock = new MockClock(new DateTime(2020, 5, 13, 15, 0, 0, 0, DateTimeKind.Utc));
+
+        private string errorResponseContent = "{\"error\": {\"code\": 404, \"message\": \"...\", \"status\": \"NOT_FOUND\"}}";
+
+        internal class MockHttpMessageHandler : HttpMessageHandler
+        {
+            private string content;
+            private HttpStatusCode statusCode;
+
+            public MockHttpMessageHandler(HttpStatusCode statusCode, string content) {
+                this.content = content;
+                this.statusCode = statusCode;
+            }
+
+            protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                HttpResponseMessage response = new HttpResponseMessage(this.statusCode)
+                {
+                    Content = new StringContent(this.content),
+                    RequestMessage = request
+                };
+                return Task.FromResult(response);
+            }
+        }
+
+        private GoogleCredential CreateSourceCredential()
+        {
+            var tokenResponse = new TokenResponse
+                {
+                    AccessToken = "ACCESS_TOKEN",
+                    ExpiresInSeconds = 60 * 10,
+                    IssuedUtc = clock.UtcNow,
+                    RefreshToken = "REFRESH_TOKEN",
+                };
+            var flowMock = new Mock<IAuthorizationCodeFlow>(MockBehavior.Strict);
+            flowMock
+                .Setup(flow => flow.Clock)
+                .Returns(clock);
+            flowMock
+                .Setup(flow => flow.AccessMethod)
+                .Returns(new BearerToken.AuthorizationHeaderAccessMethod());
+            return new GoogleCredential(new UserCredential(flowMock.Object, "my.user.id", tokenResponse));
+        }
+
+        private ImpersonatedCredential CreateImpersonatedCredentialWithIdTokenResponse()
+        {
+            var sourceCredential = CreateSourceCredential();
+            var body = new { token = OidcComputeSuccessMessageHandler.FirstCallToken };
+            var content = NewtonsoftJsonSerializer.Instance.Serialize(body);
+            var messageHandler = new MockHttpMessageHandler(HttpStatusCode.OK, content);
+            var initializer = new ImpersonatedCredential.Initializer(sourceCredential, "foo", new string[] {"scope"})
```

Comment:
Can we use "principal" instead of "foo" if that's what it means? It makes it clearer what that argument is for.

Reply:
done